### DISCUSSION TITLE
`Not Nothing` is `Something`

### DISCUSSION THREAD

**Discussion by iRon7** [PowerShell/PowerShell#24056] [General]
### Summary of the new feature / enhancement

In PowerShell `$Null` and `Nothing` (`Void`, empty array, etc.) are essential the same (see: [Enumerable null](https://learn.microsoft.com/powershell/scripting/learn/deep-dives/everything-about-null#enumerable-null))
`$Null` evaluates to "`False`":

```PowerShell
if ($Null) { 'True' } else { 'False' }
False
```

To check whether a value is holding something, the general advice is to check whether the value is not `$Null` (nothing), as in [Checking for `$null`](https://learn.microsoft.com/powershell/scripting/learn/deep-dives/everything-about-null#checking-for-null), e.g.:
```PowerShell
$Files = Get-ChildItem
if ($Null -ne $Files) {
    # Do something with $Files
}
```

Meaning, aside from counter-intuitive [Yoda condition](https://en.wikipedia.org/wiki/Yoda_conditions), it contains a [double negative](https://en.wikipedia.org/wiki/Double_negative) where (`$Null` translates to "**nothing**"" and `-ne` translates to "**is not**"):

## if `$Files` is not nothing ...

Isn't it better to use a *common positive* condition:

## if `$Files` is something (an `object`) ...

Which translates in PowerShell to:

```PowerShell
if ($Files -is [Object]) { ...
```

And to be consistent:

```PowerShell
if ($Files -isnot [Object]) { ...
```

Where the later condition contains a *single* negative as in `$Null -eq $Files` contains a *single* negative.

I aware that this is more a document issue but it is all based on the assumption that:

### `$Value -is [Object]` ⇔ `$Null -ne $Value`

But I am actually not 100% sure about that...

```PowerShell
Should -ActualValue ('abc' -is [object]) -be ($Null -ne 'abc')
Should -ActualValue ('abc' -isnot [object]) -be ($Null -eq 'abc')
Should -ActualValue (123 -is [object]) -be ($Null -ne 123)
Should -ActualValue (123 -isnot [object]) -be ($Null -eq 123)
Should -ActualValue (@() -is [object]) -be ($Null -ne @())
Should -ActualValue (@() -isnot [object]) -be ($Null -eq @())
Should -ActualValue ($Null, 1, 2 -is [object]) -be ($Null -ne $Null, 1, 2)
Should -ActualValue ($Null, 1, 2 -isnot [object]) -be ($Null -eq $Null, 1, 2)
Should -ActualValue ($Null -is [object]) -be ($Null -ne $Null)
Should -ActualValue ($Null -isnot [object]) -be ($Null -eq $Null)
Should -ActualValue ($NeverSet -is [object]) -be ($Null -ne $NeverSet)
Should -ActualValue ($NeverSet -isnot [object]) -be ($Null -eq $NeverSet)
```

One stranger in our midst is of cause the [`automationNull`](https://learn.microsoft.com/en-us/dotnet/api/system.management.automation.internal.automationnull) where these two conditions are both falsy (return nothing #24054):

```PowerShell
&{} -is [object]
&{} -isnot [object]
```

which you can't actually be compared with the `$Null` comparison:

```PowerShell
$Null -ne &{}
ParserError:
Line |
   1 |  $Null -ne &{}
     |           ~
     | You must provide a value expression following the '-ne' operator. 
```

As it requires brackets and therefore would be equal to comparing a `$Null`:

```PowerShell
Should -ActualValue ((&{}) -is [object]) -be ($Null -ne (&{}))
Should -ActualValue ((&{}) -isnot [object]) -be ($Null -eq (&{}))
```

### Proposed technical implementation details (optional)

Stimulate using `$Value -is [Object]` rather than `$Null -ne $Value`

**Comment by rhubarb-geek-nz** [comment]
> Isn't it better to use a common positive condition

How about

```
$Files = Get-ChildItem
if ($Files) {
    # Do something with $Files
}
```

That will count as true if any files were returned. Note that PowerShell will unwrap a list of one FileInfo into a single FileInfo object.

**Comment by jhoneill** [comment]
> > Isn't it better to use a common positive condition
> 
> How about
> 
> ```
> $Files = Get-ChildItem
> if ($Files) {
>     # Do something with $Files
> }
> ```
> 
> That will count as true if any files were returned. Note that PowerShell will unwrap a list of one FileInfo into a single FileInfo object.

This a standard pattern and where works were zero, empty string etc. which convert to Boolean false aren't values we want to process.    e.g. with `param ($retries) `  if zero is a valid setting,  `if ($retries)  {Set-Retries $retries}` will never set it to zero.


> In PowerShell `$Null` and `Nothing` (`Void`, empty array, etc.) are essential the same
TBH, that's not quite true;  "Never set", "Returned no objects", "Empty" all convert to Boolean false. 
But neither `'' -eq $null` nor `$null -eq '' ` return true  (empty string isn't null, and null doesn't convert) .  So there are places where we need things like `[string]::isnullOrEmpty ` but `if (-not $)`  will do the same thing if `$s` is null or holds an empty string.

**Comment by iRon7** [comment]
@rhubarb-geek-nz,

> How about

Correct, the "`$files`" example wasn't well chosen but as **jhoneil** points out, it is about a general comparison to confirm that a common `$Value` (and not just `$files`) has been set:

```PowerShell
$Value = Try { Get-Random 3 } Catch { Throw $_ } # Where `Get-Random` could be any function
if ($Value -is [object]) {                       # Which we only want to process if it contains something (including 0, "", etc)
```

@jhoneill

> TBH, that's not quite true; "Never set", "Returned no objects", "Empty" all convert to Boolean false.

Right, my statement is probably a little bluntly but TBH your statement is also not full detailed.

I think it is correct to say that they all "[*evaluate* to `$false`](https://learn.microsoft.com/powershell/module/microsoft.powershell.core/about/about_booleans#converting-from-scalar-types)" or when assigned to a variable some translate to `$Null`:

``` PowerShell
Remove-Variable NeverSet -ErrorAction SilentlyContinue
if ($Null -eq $NeverSet) { '$NeverSet is $Null' }
$NoObjects = &{ <#Returned no objects#> }
if ($Null -eq $NoObjects) { '$NoObjects is $Null' }
```

Anyways (that is not the issue), the issue/question is:
1) Can I use `$Value -is [Object]` for every `$Null -ne $Value` statement?
2) If yes, wouldn't it be better to use `$Value -is [Object]` instead of the *double-negative Yoda condition*?

**Comment by jhoneill** [comment]
> Anyways (that is not the issue), the issue/question is:
> 
> 1. Can I use `$Value -is [Object]` for every `$Null -ne $Value` statement?
> 2. If yes, wouldn't it be better to use `$Value -is [Object]` instead of the _double-negative Yoda condition_?

`$Value -is [Object]`  will fail in the case that $Value is an empty array,  there are cases where we do 
```
$stuff = @() 
$stuff += Get-Things @params1
$stuff +=Get-Things @params2
```
(In this case `$null -eq $stuff` gives $false, and `$stuff -eq $null` gives an empty array which evaluates to Boolean false)
We can test `if ($stuff)`,  or `if ($stuff.count)`, or `if ($stuff.count -gt 0 )`  Since PowerShell returns count =1 for scalars (even zero, false or empty string) you could use .`count -gt 0` as a universal test. 
    

We use double negatives or "is not what we do not want"  quite commonly 
  `$x -ne ''`  ;  `$x.count -ne 0`  etc and usually we can (and do) write `if ($value)` 

Given that it  `if ($x)`  , `if ($x -ne $null)` or the array-safe `if ($null -ne $x)`  are well understood I would avoid creating a new pattern in code to be shared with others, because it requires people to know that things that aren't object are descended from object and will make them ask why you are using a different test 

I would also say that where we need to act on the thing being tested to raise an error the code will often be clearer if we dodge the double negative. 

```
if ($x)  {
   do a lot of stuff that takes a whole page
}
else {report error}
```
Means when we read the else we have forgotten what the if was 

```
if (-not $x)  {report error}
else {
   do a lot of stuff that takes a whole page
}
```
Is a better pattern.

**Comment by rhubarb-geek-nz** [comment]
I am not sure what the issue that is being raised is. Is it about other people's coding styles?

If you were looking for a strictly typed language you've come to the wrong place. PowerShell uses Shrödinger's types. You don't know what you have until you look at it, and just because you got a list yesterday it does not mean it will be a list today. Just because it superficially appears to behave like a list does not mean it is a list.

You can still use Clean Code principles when writing PowerShell. You can get rid of a lot of if/else handling in the cases above by doing

```
if (-not $foo.DoesFrobinate) { throw 'does not frobinicate' }
if (-not $foo.HasSymbol('baz')) { throw 'no symbols to reorder' }
...do your processing here after the validation...
```

You could even use a list of scriptblocks to enumerate over to perform your validation. ( with one scriptblock per validation rule)

**Comment by mklement0** [comment]
* I personally think that the double negative isn't problematic, and from the standpoint of conceptual clarity `$null -ne $someValue` strikes me as preferable to `$someValue -is [object]` - but the two should be equivalent.

* The "Yoda syntax" is somewhat problematic, but not specific to this case: it is an unavoidable consequence of how _comparison operators_ work in PowerShell.

For extending the logic to "is this not $null, the empty string, or an empty collection?", the simplest (though not clearest) formulation is probably:

```powershell
if ($someValue.Length) { 'is a nonempty string, a non-string scalar, or a non-empty collection' }
```

This relies on the intrinsic `.Length` property rather than `.Count`, which are equivalent. While `.Count` may be conceptually preferable, using `.Length` ensures that the test also works with _strings_.

(I'm using "collection" loosely to refer to any instance that PowerShell auto-enumerates in the pipeline; for `IEnumerable` instances that aren't also `IList`, e.g. in LINQ expressions such as `([Linq.Enumerable]::Take[int]([int[]] (0, 1, 2), 2)).Length`, `.Length` triggers enumeration and therefore member-access enumeration, so it is applied to each enumerated element; while inefficient, the test should still work as intended).

**Comment by SeeminglyScience** [comment]
I'm going to convert this to a discussion since as far as I can tell there's nothing actionable here. Also might be a good question to pose to the [PowerShell discord](https://aka.ms/psdiscord).

My personal opinion is even if `-is [object]` is consistent (which I don't know the answer to) it's significantly less readable. Partially because I then have to ask if they're using it as a null check or if they have a large misconception about the type system, and partially because when framing it as a reverse null check it forces flipped contexts a bit (e.g. "oh it's used to check for null. But oh right opposite, so `is not null`". Just extra cognitive work imo)

**Comment by mklement0** [comment]
@rhubarb-geek-nz 

This issue is about developing PowerShell idioms / best practices, and should consensus be reached that a new recommendation is called for, an issue could be opened in https://github.com/PoshCode/PowerShellPracticeAndStyle/issues and/or the docs could be updated.

> If you were looking for a strictly typed language you've come to the wrong place.

PowerShell is strongly typed, but very flexible with its automatic type conversions, which is mostly a blessing, sometimes a pitfall. Knowing the conversion rules avoids the latter.

> and just because you got a list yesterday it does not mean it will be a list today.

You've expressed this sentiment repeatedly, but it is based on a misconception of the design intent behind the PowerShell pipeline:

It is meant to convey an open-ended stream of objects (0, 1, or more) and in the pipeline _there is no concept of an array or list_: objects are processed _one by one_, and only when you _collect_ pipeline output do arrays come into play, and you situationally get `[System.Management.Automation.Internal.AutomationNull]::Value` (0 objects), a single object as-is (1 object), or - for two or more objects -  an `[object[]]` array.
To ease the pain of not knowing whether collected output is an array or not, you can use `@(...)` to force the output to an array, and you can alternatively use the intrinsic `.Count` aka `.Length` properties and the intrinsic scalar indexing for unified handling of scalars and enumerables.

That, along with knowing that PowerShell _auto-enumerates_ (most) enumerables when outputting to the pipeline, should take the guesswork out of how enumerables are handled and when collection types are lost.

---

In general, I advise against repeatedly complaining about PowerShell's fundamentals in incidental remarks - it just isn't the place.

**Reply by rhubarb-geek-nz** [reply]
> _there is no concept of an array or list

Unless you mean the [Collection](https://learn.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.collection-1?view=net-8.0) implementing [IList](https://learn.microsoft.com/en-us/dotnet/api/system.collections.ilist?view=net-8.0) returned by [PowerShell.Invoke](https://learn.microsoft.com/en-us/dotnet/api/system.management.automation.powershell.invoke?view=powershellsdk-7.4.0#system-management-automation-powershell-invoke) of course.

The act of assignment assumes you have waited for all results.

Fortunately you don't have to play the 0,1,n dance that you recommend. When expecting n items to be returned simply use

```
$list = $null
$null = Get-AnyNumberOfThings -OutVariable list
```

And get an ArrayList. No need to have three separate code paths in order to handle n items.

**Reply by mklement0** [reply]
> Unless you mean the [Collection](https://learn.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.collection-1?view=net-8.0) implementing [IList](https://learn.microsoft.com/en-us/dotnet/api/system.collections.ilist?view=net-8.0) returned by [PowerShell.Invoke](https://learn.microsoft.com/en-us/dotnet/api/system.management.automation.powershell.invoke?view=powershellsdk-7.4.0#system-management-automation-powershell-invoke) of course.

That's the PowerShell _SDK_, meant (primarily) to be consumed from _outside_ PowerShell.


> The act of assignment assumes you have waited for all results.

Yes, as is enclosure in `(...)`, `$(...)` and `@(...)`; it is then that two or more objects of necessity require a container (list) of sorts, which PowerShell chose to be `[object[]]`.

> Fortunately you don't have to play the 0,1,n dance that you recommend.

You can technically avoid it, but to me what you're proposing is an even more awkward dance that - in my estimation - should never have seen the light of day:

* #3154
* #3773

> No need to have three separate code paths

You typically only need _one_ code path with direct assignment - thanks to the unified handling of scalars and lists as well as member-access enumeration.

**Reply by rhubarb-geek-nz** [reply]
> You typically only need _one_ code path with direct assignment - thanks to the unified handling of scalars and lists as well as member-access enumeration.

This approach really _does not work_ when, for example creating JSON of standard format to exchange with 3rd parties. If the JSON format defines a particular element to be an array, it needs to be an array, Not a single object because PowerShell decided to unwind it. When dealing with these data formats, if it defines a list it must be a list even if includes 0,1,or n elements. This unified approach also fails when dealing with N dimension arrays. If there is only one record but it is an array, you can't tell the difference between a response with multiple elements, or one element that is an array. Perhaps these are not the common scenarios considered by the PowerShell community, but they are real world problems.

**Reply by rhubarb-geek-nz** [reply]
I used to used create a list with New-Object then use ForEach-Object to add to the list, but the >=7.3 AMSI feature made that an unsatisfactory solution. Given [OutVariable](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_commonparameters?view=powershell-7.4#-outvariable) is frowned upon, I now have [ConvertTo-List](https://github.com/rhubarb-geek-nz/Joinery/blob/main/Joinery/ConvertToList.cs) for those scenarios where you _really need a list_.

```
$list = 1 | ConvertTo-List
Assert { $list -is [System.Collections.ArrayList] }
Assert { $list.Count -eq 1 }
$list = 1,2,3 | ConvertTo-List
Assert { $list -is [System.Collections.ArrayList] }
Assert { $list.Count -eq 3 }
$list = $null | Where-Object { $False } | ConvertTo-List
Assert { $list -is [System.Collections.ArrayList] }
Assert { $list.Count -eq 0 }
```

**Reply by iRon7** [reply]
@rhubarb-geek-nz,

I lost you, does this still apply to the original discussion (originated from an *issue*) *about developing PowerShell idioms / best practices* specifically where I question whether I might could/might/should use `$Value -is [Object]` for every `$Null -ne $Value` statement?

If this is not the case, I recommend to create a new issue (or discussion).

**Reply by rhubarb-geek-nz** [reply]
> @rhubarb-geek-nz,
> 
> I lost you, does this still apply to the original discussion (originated from an _issue_) _about developing PowerShell idioms

Yes, I was wanting to know how to handle the response where there may be 0,1 or n items just like your original discussion. I have now concluded that there is no simple way of determining if (a) you were given a single null or no records, or (b) determine if you were given multiple items, or a single item that was a collection.

**Reply by mklement0** [reply]
@rhubarb-geek-nz: 

The original issue/discussion is about an _abstract_ test: whether _something_ was returned, i.e. about  none vs. any object, irrespective of how many other objects the latter contains / enumerates.

Thus, I do think that your concern is a separate discussion; I'll provide some final brief pointers here, but if you want to continue the conversation I suggest opening a separate discussion:

* Re (a): `$null -eq $value -and $value -is [psobject]` is only true for `[System.Management.Automation.Internal.AutomationNull]::Value`, not also for `$null` - this is regrettably obscure; a green-lit, not yet implemented proposal is to support `$value -is [System.Management.Automation.AutomationNull]` - see #13465

* Re (b): In the context of a variable assignment you indeed cannot distinguish these two cases, but you can in the pipeline. If you really need this distinction, you can do something like `$isSingleOutputObjectList = (Some-Command | Tee-Object -Variable result | Measure-Object).Count -eq 1 -and $result -is [System.Collections.IList]` (note that `Tee-Object`'s `-Variable` - unlike `-OutVariable` - properly mirrors what you'd get with a direct assignment). To prevent auto-enumeration, use `Write-Output -NoEnumerate` or the unary `,` trick (`, @(1, 2) | ...`); ditto with `ConvertFrom-Json`; to make `ConvertTo-Json` treat a scalar as a an array, use `-AsArray`.

**Comment by iRon7** [comment]
@SeeminglyScience

> I'm going to convert this to a discussion since as far as I can tell there's nothing actionable here.

Thanks, that was the right thing to do for this issue, I am not really a Discord user and I fear that it might be even more difficult there to keep the focus on the actual question(s) in the issue.
I appreciate you opinion.

@mklement0

> ... `$someValue -is [object]` - but the two should be equivalent

Thanks for the confirmation and I appreciate you opinion too.
Yet, I personally don't like the "intrinsic `.Length` property" alternative as it might cause an error in strict mode for "*a non-string scalar*":

```PowerShell
Set-StrictMode -Version Latest
$someValue = 3
if ($someValue.Length) { 'is a nonempty string, a non-string scalar, or a non-empty collection' }

PropertyNotFoundException: The property 'Length' cannot be found on this object. Verify that the property exists.
```

**Reply by mklement0** [reply]
Yes, that is unfortunate and should be considered a bug:

* #2798

It is the reason I personally only ever use at most `-Version 1` of  `Set-StrictMode`.

In practice, there is a lot of code that will break even if `-Version 1` is in effect, so it's crucial to carefully manage strict mode to ensure that it applies only to _your_ code (which is less of a problem inside modules).

The fact that the bug hasn't been addressed since it was first reported in 2016 is another indicator that `Set-StrictMode` isn't a priority.
The PowerShell extension for Visual Studio Code is a design-time alternative.
